File: 0948-sort-an-array/0948-sort-an-array.java
class Solution {
    public int[] sortArray(int[] nums) {
        if (nums.length == 0) return nums;

        int min = Integer.MAX_VALUE, max = Integer.MIN_VALUE;
        
        // Find min and max
        for (int num : nums) {
            min = Math.min(min, num);
            max = Math.max(max, num);
        }

        int range = max - min + 1;
        int[] freq = new int[range];

        // Count frequency
        for (int num : nums) {
            freq[num - min]++;
        }

        // Fill sorted result
        int index = 0;
        for (int i = 0; i < range; i++) {
            while (freq[i]-- > 0) {
                nums[index++] = i + min;
            }
        }

        return nums;
    }
}

File: 0948-sort-an-array/0948-sort-an-array.rs
impl Solution {
    pub fn sort_array(nums: Vec<i32>) -> Vec<i32> {
        if nums.is_empty() {
            return vec![];
        }

        let min_val = *nums.iter().min().unwrap();
        let max_val = *nums.iter().max().unwrap();
        let range = (max_val - min_val + 1) as usize;

        let mut freq = vec![0; range];

        for &num in &nums {
            freq[(num - min_val) as usize] += 1;
        }

        let mut sorted = Vec::with_capacity(nums.len());
        for (i, &count) in freq.iter().enumerate() {
            for _ in 0..count {
                sorted.push(i as i32 + min_val);
            }
        }

        sorted
    }
}
